Add CI job that validates and lints vimscript

alerque opened the pull request:
1. Configure [`vint`](https://github.com/Kuniwak/vint) so you can run it locally on this project (`vint .` from the root).
2. Fix bugs found by linter including...
    * An issue that would trip up some functions if any double-byte encoded text was present in the buffer.
    * A bunch of issues that would only have turned up if people had remapped standard keys. For example this is now compatible with `vim-smartbraces` again.
    * A few issues that would only have turned up depending on user RC settings. It should be a lot more robust now.
3. Fix all the warnings about bad vimscript practices that it turned up.
4. Add a CI job using GitHub Actions that will run the linter on pull requests.

For the latter to work somebody who is an owner in this org will need to [signup for Actions](https://github.com/features/actions/signup/?account=vim-pandoc) and enable it for this org. I don't have the correct permissions. Additionally after signing up and enabling it for this org please note that the CI job will not start running until after this PR is merged because a workflow configuration has to exist in `master` before it considers running it on PRs. You can see the CI job in action [on my fork](https://github.com/alerque/vim-pandoc/actions) where I tested it.

Please also note that this **should be put through the passes** before merging to make sure I didn't break anything significant. I tried some of my basic usage but I don't use half the functions here so I wouldn't know if I broke them.

alerque commented:
(**Please don't squash this!** A regular merge would be better because one of these commits is much more likely to turn up in a `git blame` than the others and it will be a lot easier to figure out what actually broke if they are separate. One only fixes hundreds of trivial warnings, the other makes a handful of substantial changes.)

fmoralesc commented:
Cool! I will signup for Actions, and then we can move forward.

 It would be nice to split the commit that fixes bugs into separate things, or at least to have a description of the bugs in the commit message.

alerque commented:
I can split it pretty easily the next time I'm at a computer.

fmoralesc commented:
Perfect, then ;) Can I repeat how good an idea this PR is??

alerque commented:
I split up the "bug fixes" commit into logical groupings by the kind of issue they were fixing, with some explanations.

I _could_ also split up the lint warnings one but it would take an hour or so to sort out because in many cases more than one type of fix appears in the same line of code, so I would have to apply most of them from scratch in order by type separately. I don't think that's going to be worth it because even split by type it won't be much more useful on a `git blame`. Any breakage almost certainly has to do with my guessing which comparisons were case sensitive and which not, and if that ever comes up I would debug it by doing a hunk-by-hunk restaging of the commit. Having the different lint warning types split out won't make that appreciably easier. I did add some comments though.

alerque commented on autoload/pandoc/toc.vim:
```diff
@@ -59,33 +60,33 @@ function! pandoc#toc#Update()
     set modifiable
-    silent %s/\v^([^|]*\|){2,2} #//e
-    for l in range(1, line("$"))
+    call substitute('%', '\v^([^|]*\|){2,2} #', '')
```
Of all the changes here, this is the one I am least confident of. Using `:s///` is subject to all kinds of user mode settings, and the `substitute()` command eliminates those but it also handles magic modes differently (`\v`) and I wasn't sure how to actually test that this is still doing what it was supposed to.

fmoralesc commented:
> I split up the "bug fixes" commit into logical groupings by the kind of issue they were fixing, with some explanations. 

Thanks! I took a look and everything looks in order ;) 

Indeed, I don't think there is much benefit in splitting the lint fixes.

alerque commented:
Maybe have a quick look at 9897435 and see if you know how to confirm that it is working as expected. If so I'm good for merge (so squash!).

alerque commented:
That last force push was just to fix a typo in one of the commit messages, no serious harm ;-)

fmoralesc commented:
Perfect, I think we are ready to go. Just to confirm, we should just merge, not squash and merge, right?

alerque commented:
If you want my opinion Github should just squish the squash option altogether and everybody should make [`git revise`](https://github.com/mystor/git-revise) part of their minute-by-minute workflow. Given his [squash-happy habits](https://github.com/liuchengxu/vim-clap/pull/83) that would leave developers like @liuchengxu a bit high and dry, but then he might actually have a reason to work on [my feature request](https://github.com/liuchengxu/vim-clap/issues/64)!
